PMM-7925: PMM regular user server deployment using Podman [Technical Preview] (#827)
percona/pmm-doc: docs/setting-up/server/podman.md

    ```sh
    systemctl --user stop pmm-server
    ```

2. Backup the container image.

    ```sh
    podman rename pmm-server pmm-server-backup
    podman tag pmm-server-backup pmm-server-backup:2.28.0
    ```

    !!! caution alert alert-warning "Important"
        Change X.Y.Z (2.28.0) to PMM version you are running, or version of your choice that you could later restore from

3. Backup the data.

    ```sh
    podman volume export pmm-server --output pmm-server-backup.tar
    ```

    !!! caution alert alert-warning "Important"
        If you changed default name by `PMM_VOLUME_NAME` environment variable, use that name after `export` instead of `pmm-server` (which is default volume name).

## Upgrade

!!! summary alert alert-info "Summary"
    - Perform a backup.
    - Update PMM tag.
    - Pre-pull image.
    - Run it.

---

!!! caution alert alert-warning "Important"
    Downgrades are not possible. To go back to using a previous version you must have created a backup of it before upgrading.

!!! hint alert alert-success "Tip"
    To see what release you are running, use the *PMM Upgrade* panel on the *Home Dashboard*, or run:

```sh
podman exec -it pmm-server \

> curl -ku admin:admin https://localhost/v1/version

Reviewer: Why is this being done with `podman exec`? Just to avoid saying, "install curl"?

Author: port mapping was in my mind probably. That command would succeed inside. Outside the container you need alsp mention that port need to be correct.

    Restart=on-failure
    RestartSec=20

    [Install]
    Alias=%N
    WantedBy=default.target

    EOF
    ```

    Create `~/.config/pmm-server/pmm-server.env` file:

    ```sh
    mkdir -p ~/.config/pmm-server/
    cat << "EOF" > ~/.config/pmm-server/pmm-server.env
    # env file passed to the container
    # full list of environment variables:
    # https://www.percona.com/doc/percona-monitoring-and-management/2.x/setting-up/server/docker.html#environment-variables

    # keep updates disabled
    # do image replacement instead (update the tag and restart the service)
    DISABLE_UPDATES=1

    # Enable DBaaS feature
    #ENABLE_DBAAS=1
    EOF
    ```

2. Configure.

    SystemD service passes the environment parameters from the `pmm-server.env `file (in `~/.config/pmm-server/pmm-server.env`) to PMM. For more information about container environment variables, check [Docker Environment].

    SystemD service uses some environment variables that could be customized if needed:

    ```text
    Environment=PMM_PUBLIC_PORT=8443
    Environment=PMM_VOLUME_NAME=%N
    Environment=PMM_TAG=2.29.0
    Environment=PMM_IMAGE=docker.io/percona/pmm-server
    ```

>     You can override the environment variables by defining them in the file  `~/.config/pmm-server/env`. For example, to override the path to a custom registry `~/.config/pmm-server/env`:

Author: these are 2 different:
- `pmm-server.env` are envs for PMM server
- `env` are envs for systemd service

>     You can override the environment variables by defining them in the file  `~/.config/pmm-server/env`. For example, to override the path to a custom registry `~/.config/pmm-server/env`:

Author: we probably need to be more clear about it. I could add another improvement PR, after merging this one.

>     You can override the environment variables by defining them in the file  `~/.config/pmm-server/env`. For example, to override the path to a custom registry `~/.config/pmm-server/env`:

Author: pushed the description

>     You can override the environment variables by defining them in the file  `~/.config/pmm-server/env`. For example, to override the path to a custom registry `~/.config/pmm-server/env`:

Reviewer 2: okay I see, thank you makes sense

    ```sh
    mkdir -p ~/.config/pmm-server/
    cat << "EOF" > ~/.config/pmm-server/env
    PMM_TAG=2.28.0
    PMM_IMAGE=docker.io/percona/pmm-server
    PMM_PUBLIC_PORT=8443
    EOF
    ```

    !!! caution alert alert-warning "Important"
        Ensure that you modify PMM_TAG in `~/.config/pmm-server/env` and update it regularly as Percona cannot update it. It needs to be done by you.

3. Enable and Start.

    ```sh
    systemctl --user enable --now pmm-server
    ```

> 4. Visit `https://localhost:8443` to see the PMM user interface in a web browser. (If you are accessing host remotely, replace `localhost` with the IP or server name of the host.)

Author: ` --now` in `enable` starts the service, that was Ceri's suggestion, so I removed separate start.

> 4. Visit `https://localhost:8443` to see the PMM user interface in a web browser. (If you are accessing host remotely, replace `localhost` with the IP or server name of the host.)

Reviewer 2: okay, makes sense

    - Pre-pull image.
    - Run it.

---

!!! caution alert alert-warning "Important"
    Downgrades are not possible. To go back to using a previous version you must have created a backup of it before upgrading.

!!! hint alert alert-success "Tip"
    To see what release you are running, use the *PMM Upgrade* panel on the *Home Dashboard*, or run:

    ```sh
    podman exec -it pmm-server \
    curl -ku admin:admin https://localhost/v1/version
    ```

(If you are accessing the podman host remotely, replace `localhost` with the IP or server name of the host.)

1. Perform a [backup](#backup).

2. Update PMM tag.

    Edit `~/.config/pmm-server/env` and create/update with a new tag from [latest release](https://per.co.na/pmm/latest):

    ```sh
    sed -i "s/PMM_TAG=.*/PMM_TAG=2.29.0/g" ~/.config/pmm-server/env
    ```

3. Pre-pull image for faster restart.

    <div hidden>
    ```sh
    sed -i "s/PMM_TAG=.*/PMM_TAG=2.29.0-rc/g" ~/.config/pmm-server/env
    sed -i "s|PMM_IMAGE=.*|PMM_IMAGE=docker.io/perconalab/pmm-server|g" ~/.config/pmm-server/env
    ```
    </div>

    ```sh
    source ~/.config/pmm-server/env
    podman pull ${PMM_IMAGE}:${PMM_TAG}
    ```

> 4. Run it.

Reviewer 3: ```suggestion
4. Run PMM.
```

    ```sh
    systemctl --user restart pmm-server
    ```

<div hidden>
```sh
sleep 30
timeout 60 podman wait --condition=running pmm-server
```
</div>

## Restore

!!! summary alert alert-info "Summary"
    - Stop PMM server.

>     - Run on previous image.

Reviewer 3: ```suggestion
    - Run PMM on the previous image.
```

    - Restore the volume.
    - Start PMM Server.

---

!!! caution alert alert-warning "Important"
    You must have a [backup](#backup) to restore from.
    You need to perform restore only if you have issues with upgrade or with the data.

1. Stop PMM server.

    ```sh
    systemctl --user stop pmm-server
    ```

> 2. Run on previous image.

Reviewer 3: ```suggestion
2. Run PMM on the previous image.
```

    Edit `~/.config/pmm-server/env` file:

    ```sh
    sed -i "s/PMM_TAG=.*/PMM_TAG=2.28.0/g" ~/.config/pmm-server/env
    ```

    !!! caution alert alert-warning "Important"
        X.Y.Z (2.28.0) is the version you used before upgrade and you made Backup with it

3. Restore the volume.

    ```sh
    podman volume import pmm-server pmm-server-backup.tar
    ```

4. Start PMM Server.

    ```sh
    systemctl --user start pmm-server
    ```

    <div hidden>
    sleep 30
    timeout 60 podman wait --condition=running pmm-server
    ```
    </div>

## Remove

!!! summary alert alert-info "Summary"
    - Stop PMM server.
    - Remove (delete) volume.
    - Remove (delete) images.

---

!!! caution alert alert-warning "Caution"

>     These steps delete the PMM Server Docker image and any accumulated PMM metrics data.

Reviewer 3: ```suggestion
    These steps delete the PMM Server Docker image and the associated PMM metrics data.
```

1. Stop PMM server.

    ```sh
    systemctl --user stop pmm-server
    ```

2. Remove volume.

    <div hidden>
    ```sh
    #wait for container to stop
    podman wait --condition=stopped pmm-server || true
    sleep 10
    ```
    </div>

    ```sh
    podman volume rm --force pmm-server
    ```

> 3. Remove the images.

Reviewer 3: ```suggestion
3. Remove the Podman images.
```

# Podman

!!! caution alert alert-warning "Caution"
    Podman support is currently in [technical preview](../../details/glossary.md#technical-preview) and is subject to change.

How to run PMM Server with Podman on our [Docker image].

!!! note alert alert-primary ""

>     The tags used here are for the current release. Other [tags] are available.

Author: good suggestion! BTW @rasika-chivate - we should consider to use macros in documentation, so it will always points to the right version.

    # optional env file that could override previous env settings for this unit
    EnvironmentFile=-%h/.config/pmm-server/env

    ExecStart=/usr/bin/podman run --rm --replace=true --name=%N -p ${PMM_PUBLIC_PORT}:443/tcp --ulimit=host --volume=${PMM_VOLUME_NAME}:/srv --env-file=${PMM_ENV_FILE} --health-cmd=none --health-interval=disable ${PMM_IMAGE}:${PMM_TAG}
    ExecStop=/usr/bin/podman stop -t 10 %N
    Restart=on-failure
    RestartSec=20

    [Install]
    Alias=%N
    WantedBy=default.target

    EOF
    ```

    Create `~/.config/pmm-server/pmm-server.env` file:

    ```sh
    mkdir -p ~/.config/pmm-server/
    cat << "EOF" > ~/.config/pmm-server/pmm-server.env
    # env file passed to the container
    # full list of environment variables:
    # https://www.percona.com/doc/percona-monitoring-and-management/2.x/setting-up/server/docker.html#environment-variables

    # keep updates disabled
    # do image replacement instead (update the tag and restart the service)
    DISABLE_UPDATES=1

    # Enable DBaaS feature
    #ENABLE_DBAAS=1
    EOF
    ```

    Enable PMM sever service to start with user:

    ```sh

>     systemctl --user enable pmm-server

Reviewer: @denisok is this supposed to start it too, as that would be:
```sh
systemctl --user enable --now pmm-server
```
